### PR TITLE
Handle Netty channel connection failures

### PR DESCRIPTION
**Pull request by injectives**
This update fixes a bug where connection acquisition may hang due to a failure to complete `ChannelPromise` of a `FailedChannel`.

### CODE REVIEW

File: driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/NettyConnectionProvider.java [hunk 1]
```diff
@@ -111,42 +110,29 @@ public CompletionStage<Connection> acquireConnection(
...
-        });
-        return future;
+        return installChannelConnectedListeners(address, bootstrap.connect(socketAddress), connectTimeoutMillis)
+                .thenApply(channel -> {
+                    // remove timeout handler from the pipeline once TLS and Bolt handshakes are
+                    // completed. regular protocol
+                    // messages will flow next and we do not want to have read timeout for them
+                    channel.pipeline().remove(ConnectTimeoutHandler.class);
+                    return channel;
+                })
```
Comment (robsdedude) [inline driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/NettyConnectionProvider.java]: Nit: should/could this live inside `installChannelConnectedListeners` as that's also the place where the timeout handler is added to the pipeline?

**Review by gjmwoods — APPROVED** [review]

File: driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/HandshakeCompletedListener.java [hunk 2]
```diff
@@ -1,80 +0,0 @@
...
-            ChannelPromise connectionInitializedPromise,
-            NotificationConfig notificationConfig,
-            Clock clock,
-            CompletableFuture<Long> latestAuthMillisFuture) {
-        requireNonNull(clock, "clock must not be null");
-        this.authMap = authMap;
-        this.userAgent = requireNonNull(userAgent);
-        this.boltAgent = requireNonNull(boltAgent);
-        this.routingContext = routingContext;
-        this.connectionInitializedPromise = requireNonNull(connectionInitializedPromise);
```
Comment (robsdedude) [inline driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/HandshakeCompletedListener.java:57]: Are we fine with dropping these NonNull guards?

File: driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/ChannelConnectedListener.java [hunk 3]
```diff
@@ -66,11 +66,11 @@ public void operationComplete(ChannelFuture future) {
                         error = new ServiceUnavailableException(
                                 String.format("Unable to write Bolt handshake to %s.", this.address), error);
                     }
-                    this.handshakeCompletedPromise.setFailure(error);
+                    this.handshakeCompletedFuture.completeExceptionally(error);
                 }
             });
         } else {
-            handshakeCompletedPromise.setFailure(databaseUnavailableError(address, future.cause()));
+            handshakeCompletedFuture.completeExceptionally(databaseUnavailableError(address, future.cause()));
```
Comment (robsdedude) [inline driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/ChannelConnectedListener.java:73]: Is it possible to craft a unit test that prevents a regression? I.e., fails with the old code, passes with the new code by simulating a failing channel. Or is there too much coupling between netty and the driver here?

File: driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/HandshakeCompletedListener.java (hunk 2 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/HandshakeCompletedListener.java:57]: Updated to keep 3 of those. Thank you.

File: driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/NettyConnectionProvider.java (hunk 1 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/NettyConnectionProvider.java]: Updated.

File: driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/ChannelConnectedListener.java (hunk 3 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/internal/bolt/basicimpl/async/connection/ChannelConnectedListener.java:73]: Updated.

**Review by robsdedude — APPROVED** [review]
:sparkler:

**Pull request merged**